### PR TITLE
prim_shared: Fix clip mask perspective interpolation.

### PR DESCRIPTION
**Pull request by emilio**
This was really hard to track down, the comment there really convinced me that
the code was correct, so I ended up looking at everything else before that...

This fixes https://bugzilla.mozilla.org/show_bug.cgi?id=1494453.

---
This change is [Reviewable](https://reviewable.io/reviews/servo/webrender/3145)

### CODE REVIEW

**Comment by emilio** [comment]
https://treeherder.mozilla.org/#/jobs?repo=try&revision=a03726cfc8516204b07957a0a38e96d84e1a48f4

File: webrender/res/prim_shared.glsl [hunk 1]
```diff
@@ -241,8 +241,7 @@ float do_clip() {
         return 1.0;
     }
     // anything outside of the mask is considered transparent
-    //Note: we assume gl_FragCoord.w == interpolated(1 / vClipMaskUv.w)
-    vec2 mask_uv = vClipMaskUv.xy * gl_FragCoord.w;
+    vec2 mask_uv = vClipMaskUv.xy / vClipMaskUv.w;
```
[the remark this answers is not in the document]
Comment (emilio) [reply webrender/res/prim_shared.glsl]: I don't know what would guarantee that given `write_vertex` has:

```glsl
    gl_Position = uTransform * vec4(final_pos, z, 1.0);
```

[the remark this answers is not in the document]
Comment (kvark) [reply webrender/res/prim_shared.glsl]: `write_vertex` also is supposed not supposed to be used for transforms that have any perspective, hence both `vClipMaskUv.w == 1.0` and `gl_FragCoord.w == 1.0`, so the assumption still stands.

[the remark this answers is not in the document]
Comment (emilio) [reply webrender/res/prim_shared.glsl]: Well, in this test-case we're definitely hitting `write_vertex`, with a `world_pos` with the `w` component different than 1. We're still writing `1.0` in `gl_Position`, but `vClipMaskUv.w == world_pos.w`, which is not one..

[the remark this answers is not in the document]
Comment (emilio) [reply webrender/res/prim_shared.glsl]: The rest of `write_vertex` also doesn't seem to assume `world_pos.w == 1.0`, fwiw.

File: webrender/res/prim_shared.glsl [hunk 2]
```diff
@@ -121,14 +121,12 @@ VertexInfo write_vertex(RectWithSize instance_rect,
...
+    vec2 device_pos = world_pos.xy * uDevicePixelRatio;
 
     // Apply offsets for the render task to get correct screen location.
-    vec2 final_pos = device_pos + snap_offset -
-                     task.content_origin +
-                     task.common_data.task_rect.p0;
+    vec2 final_offset = snap_offset - task.content_origin + task.common_data.task_rect.p0;
 
-    gl_Position = uTransform * vec4(final_pos, z, 1.0);
+    gl_Position = uTransform * vec4(device_pos + final_offset, z, world_pos.w);
```
Comment (kvark) [inline webrender/res/prim_shared.glsl]: we should also multiply `final_offset` and `z` by `world_pos.w`

Comment (emilio) [reply webrender/res/prim_shared.glsl]: Done

**Pull request merged**